debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -0,0 +1,159 @@
+        private Supplier<LoggingContext.PreviousContext> loggingContextSupplier;
+
+        public Builder<T> pollInterval(Duration pollInterval) {
+            this.pollInterval = pollInterval;
+            return this;
+        }
+
+        public Builder<T> maxQueueSize(int maxQueueSize) {
+            this.maxQueueSize = maxQueueSize;
+            return this;
+        }
+
+        public Builder<T> maxBatchSize(int maxBatchSize) {
+            this.maxBatchSize = maxBatchSize;
+            return this;
+        }
+
+        public Builder<T> loggingContextSupplier(Supplier<LoggingContext.PreviousContext> loggingContextSupplier) {
+            this.loggingContextSupplier = loggingContextSupplier;
+            return this;
+        }
+
+        public ChangeEventQueue<T> build() {
+            return new ChangeEventQueue<T>(pollInterval, maxQueueSize, maxBatchSize, loggingContextSupplier);
+        }
+    }
+
+    public void enqueue(T record) {
+        LoggingContext.PreviousContext previousContext = loggingContextSupplier.get();
+
+        try {
+            queue.put(record);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Placed source record '{}' into queue", record);
+            }
+        } catch (InterruptedException e) {
+            LOGGER.debug("received interrupt request");
+            // clear the interrupted status
+            Thread.interrupted();
+        } finally {
+            previousContext.restore();
+        }
+    }
+
+    public List<T> poll() {
+        LoggingContext.PreviousContext previousContext = loggingContextSupplier.get();
+
+        try {
+            LOGGER.debug("polling records...");
+            List<T> records = new ArrayList<>();
+            final Timer timeout = Threads.timer(Clock.SYSTEM, Temporals.max(pollInterval, ConfigurationDefaults.RETURN_CONTROL_INTERVAL));
+            while (!timeout.expired() && queue.drainTo(records, maxBatchSize) == 0) {
+                throwProducerFailureIfPresent();
+
+                try {
+                    LOGGER.debug("no records available yet, sleeping a bit...");
+                    // no records yet, so wait a bit
+                    metronome.pause();
+                    LOGGER.debug("checking for more records...");
+                } catch (InterruptedException e) {
+                    // we've been requested to stop polling
+                    Thread.interrupted();
+                    break;
+                }
+            }
+            return records;
+        } finally {
+            previousContext.restore();
+        }
+    }
+
+    public void producerFailure(final Throwable producerFailure) {
+        this.producerFailure = producerFailure;
+    }
+
+    private void throwProducerFailureIfPresent() {
+        if (producerFailure != null) {
+            throw new ConnectException("An exception ocurred in the change event producer. This connector will be stopped.", producerFailure);
```

Comment:
I'd not refer to connector term as we might use the `ChangeEventQueue` in other scenarios not only driving mani connector loop.

Reply:
Interesting point. I think we could change thise once we really use the queue for other things (anything specific coming to your mind?). I like the message as it makes clear what's going to happen. But we can generify that later on as needed.